[Depr] raise_on_error kwarg with errors kwarg in astype#14878

m-charlton opened the pull request:
- [x] closes #14878 
 - [x] tests passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Please check that the entry in `whatsnew/v0.20.0.txt`. Unsure that the update was in the 
`_whatsnew_0200.deprecations` or `_whatsnew_0200.prior_deprecations` so put it in the
former.

m-charlton committed:
astype method now takes dict mapping col names to datatypes #14761

Updating documentation to reflect change

m-charlton committed:
Updates after review

DataFrame.astype now allows changing the dtype of a column by
passing a dict mapping column name to dtype.

m-charlton committed:
Corrections to docs after review feedback

DataFrame.astype now allows setting the type of columns
by passing a dict mapping column to dtype.

m-charlton committed:
Merge branch 'update_docs_astype_with_dict_#14761' of https://github.com/m-charlton/pandas into update_docs_astype_with_dict_#14761

m-charlton committed:
Merge branch 'update_docs_astype_with_dict_#14761'

m-charlton committed:
Removing uneeded changes

Mistakenly added changes carried out in v0.19 to v0.20

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
removing stray orig file after merge

m-charlton committed:
Deprecate 'raise_on_error' kwarg with 'errors' for DataFrame.astype

Valid arguments for new 'errors' kwarg are 'ignore' or 'raise'
see #14878

jreback commented on pandas/core/generic.py:
```diff
@@ -3073,7 +3075,9 @@ def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
             the same type. Alternatively, use {col: dtype, ...}, where col is a
             column label and dtype is a numpy.dtype or Python type to cast one
             or more of the DataFrame's columns to column-specific types.
-        raise_on_error : raise on invalid input
```
leave this in here and just say DEPRECATED.

@jorisvandenbossche IIRC that is our convention?

add a versionadded tag 0.20.0 for errors

m-charlton replied:
Will do.

jorisvandenbossche replied:
yes, can leave it (but put it at the end)

jreback commented on pandas/core/generic.py:
```diff
@@ -3073,7 +3075,9 @@ def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
+        errors : {'raise', 'ignore'}, default 'raise'
+            - ``raise`` : allow exceptions to be raised on invalid input
+            - ``ignore`` : suppress raising exceptions on invalid input
         kwargs : keyword arguments to pass on to the constructor
```
I think you need a blank line before kwargs (to make the sub-list work)

m-charlton replied:
I've just checked and the sublist is rendered fine with, or without a line between `ignore`
and kwargs. I can add an extra line if that is the convention,

jreback commented on pandas/core/internals.py:
```diff
@@ -455,12 +455,11 @@ def downcast(self, dtypes=None, mgr=None):
 
-    def _astype(self, dtype, copy=False, raise_on_error=True, values=None,
+    def _astype(self, dtype, copy=False, errors='raise', values=None,
                 klass=None, mgr=None, **kwargs):
```
can you check the errors in ['raise', 'ignore'] at the beginning of the function and raise a ValueError otherwise (and add a test for this)

m-charlton replied:
Sure, there are two 'public' `astype(...)` methods: 

- `NDFrame.astype(...)` in `pandas/core/generic.py`
- `Block.astype(...)` in `pandas/core/internals.py`

In addition there is a 'protected' `Block._astype(...)` method in  
`pandas/core/internals.py`. Should I only put the checks in the 'public' 
methods?

Bearing in mind that the `raise_on_error` kwarg is going to be deprecated for
`DataFrame.where()` and replaced with the `errors` kwarg it would make sense to
put the code that checks the validity of the arguments in one place. Do we have 
any existing code where we put such validity checking functions/methods?

I notice that both `NDFrame` & `Block` inherit from `PandasObject` but, I'm not 
sure that this is the correct thing to do. Should I put a function in 
`pandas/core/base.py`?

jreback replied:
Block is completely internal
u can put the check there

we have centralized checks but no need in this case 

out in _astype as that's where it's actually used 
parameter validation is best down where it's actually

m-charlton committed:
Check & unit test added for validity of arguments to 'errors' kwarg. Docstrings
to astype clarified.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -3073,7 +3075,14 @@ def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
+            instead
+        errors : {'raise', 'ignore'}, default 'raise'
+            - ``raise`` : allow exceptions to be raised on invalid input
+            - ``ignore`` : suppress raising exceptions on invalid input
```
maybe specify what happens of there is an error? (original is returned)

jorisvandenbossche commented on pandas/core/internals.py:
```diff
@@ -455,17 +455,23 @@ def downcast(self, dtypes=None, mgr=None):
+
+        if errors not in errors_legal_values:
+            invalid_arg = "Expected value of kwarg 'errors' to be one of %s. "\
+                "Supplied value is '%s'" % (', '.join("'%s'" % arg for arg in
```
style comment: can you do the string continuation with putting `(  )` around it instead of the `\`. And can you use ´.format(..)´ instead of `%` (can be on a separate line)

m-charlton replied:
Will do. I'll change the output message to:

```Expected value of kwarg 'errors' to be one of ['raise', 'ignore']. Supplied value 'True'```

This will do away with the need for the messy formatting.

jreback replied:
@m-charlton can you change this to the {} formatting syntax?

m-charlton replied:
Will do

jorisvandenbossche commented on pandas/tests/test_internals.py:
```diff
@@ -566,6 +566,13 @@ def test_astype(self):
                 self.assertEqual(tmgr.get('d').dtype.type, t)
 
+    def test_illegal_arg_for_errors_in_astype(self):
+        """ ValueError exception raised when illegal value used for errors """
```
Can you turn this into a normal comment (with `#`)? Otherwise the test name does not show up in the nosetests output. 
You can also add the issue number

jorisvandenbossche replied:
I would also move the test to the high-level `astype` tests for frame/series

m-charlton replied:
will do

m-charlton replied:
I've looked at the existing test cases in `pandas/tests/series` & 
`pandas/tests/frame` and can't find a natural place to put this test.

I was thinking of adding a new file called say `test_astype.py` in
`pandas/tests/frame`, as there is an existing `test_apply.py`

jreback replied:
no just grep for astype and you will see lots of tests

jreback replied:
ok this test is fine here.

jorisvandenbossche replied:
@m-charlton astype tests are located in pandas/tests/frame/test_dtypes.py

m-charlton replied:
Will do

m-charlton commented:
Update after review

jreback commented on pandas/core/internals.py:
```diff
@@ -455,17 +455,23 @@ def downcast(self, dtypes=None, mgr=None):
         """
+        errors_legal_values = ('raise', 'ignore')
+
+        if errors not in errors_legal_values:
```
just do this inline here

``if errors not in ['raise', 'ignore']:``

jreback reviewed (changes requested):
small error message formatting change. ping when pushed and green.

jorisvandenbossche commented:
xref https://github.com/pandas-dev/pandas/pull/14877 we should try to clarify that the `errors` keyword only is meant for errors during the type conversion. Maybe we can change the "invalid input" to something else? Eg "invalid data for the provided dtype"?

m-charlton commented:
Back from Christmas/New Year break. Will makes changes today and pushh for review

m-charlton committed:
Update after code review #1478

Tests added for deprected 'raise_on_error' kwarg & new 'errors' kwarg.
Clarified docstring for DataFrame.astype method

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -523,6 +523,24 @@ def test_timedeltas(self):
         result = df.get_dtype_counts().sort_values()
         assert_series_equal(result, expected)
 
+    def test_illegal_arg_for_errors_in_astype(self):
```
you can make this a single test.

as the same test in tests/series/test_dtypes.py as well

jreback commented:
minor test change. ping on green.

m-charlton committed:
Update after code review of unit tests #1478

Unit tests for DataFrame.astype merged. Dupliacted those tests for
Series.astype. Both testing deprecation of 'raise_on_error' kwarg.

jorisvandenbossche committed:
switch kwarg order in docstring

jorisvandenbossche commented:
@m-charlton Thanks a lot!
